File: Algorithms/0063_Unique_Paths_II/Java/UniquePathsIiSolution1.java

// Space: O(n)
// Time: O(n)

public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid[0][0] == 1) return 0;
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        int[][] tempBoard = new int[m][n];
        for (int i = 0; i < tempBoard.length; i++) {
            Arrays.fill(tempBoard[i], 0);
        }
        tempBoard[0][0] = 1;
        for (int y = 0; y < m; y++) {
            for (int x = 0; x < n; x++) {
                if (obstacleGrid[y][x] != 1) {
                    tempBoard[y][x] += y - 1 >= 0 ? tempBoard[y - 1][x] : 0;
                    tempBoard[y][x] += x - 1 >= 0 ? tempBoard[y][x - 1] : 0;
                }
            }
        }
        return tempBoard[m - 1][n - 1];
    }
}

File: Algorithms/0063_Unique_Paths_II/Python/Unique_Paths_II_Solution_1.py

# Space: O(n)
# Time: O(n)


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid):
        if obstacleGrid[0][0] == 1: return 0
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])

        temp_board = [[0 for _ in range(n)] for _ in range(m)]
        temp_board[0][0] = 1

        for y in range(m):
            for x in range(n):
                if obstacleGrid[y][x] != 1:
                    temp_board[y][x] += temp_board[y - 1][x] if y - 1 >= 0 else 0
                    temp_board[y][x] += temp_board[y][x - 1] if x - 1 >= 0 else 0

        return temp_board[m - 1][n - 1]




